internal/states/encryption/encryption.go

```diff
@@ -0,0 +1,271 @@
+	//
+	// - If the user provided no configuration, this function must return a flow that passes through the data
+	//   unmodified.
+	// - The default configuration is always ignored in this case because it is only the default for remote states.
+	// - If the user provided a non-default HCL or environment configuration, these configurations should be merged
+	//   with the environment taking precedence. The default configuration should be ignored.
+	//
+	// Tip: encryptionconfig.ConfigMap.Merge implements these precedence rules.
+	PlanFile() (encryptionflow.PlanFlow, error)
+}
+
+type encryption struct {
+	encryptionConfigs         encryptionconfig.ConfigMap
+	decryptionFallbackConfigs encryptionconfig.ConfigMap
+	mutex                     sync.Mutex
+	logger                    hclog.Logger
+}
+
+func (e *encryption) ApplyEnvConfigurations(
+	encryption map[encryptionconfig.Key]encryptionconfig.Config,
+	decryptionFallback map[encryptionconfig.Key]encryptionconfig.Config,
+) error {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	for key, config := range encryption {
+		if err := key.Validate(); err != nil {
+			return fmt.Errorf("failed to parse encryption configuration from environment (%w)", err)
+		}
+		meta := encryptionconfig.Meta{
+			encryptionconfig.SourceEnv,
+			key,
+		}
+		e.encryptionConfigs[meta] = config
+	}
+
+	for key, config := range decryptionFallback {
+		if err := key.Validate(); err != nil {
+			return fmt.Errorf("failed to parse decryption fallback configuration from environment (%w)", err)
+		}
+		e.decryptionFallbackConfigs[encryptionconfig.Meta{
+			encryptionconfig.SourceEnv,
+			key,
+		}] = config
+	}
+
+	return nil
+}
+
+func (e *encryption) ApplyHCLEncryptionConfiguration(key encryptionconfig.Key, config encryptionconfig.Config) error {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	if err := key.Validate(); err != nil {
+		return fmt.Errorf("failed to parse encryption configuration from HCL (%w)", err)
+	}
+	meta := encryptionconfig.Meta{
+		encryptionconfig.SourceHCL,
+		key,
+	}
+	e.encryptionConfigs[meta] = config
+	return nil
+}
+
+func (e *encryption) ApplyHCLDecryptionFallbackConfiguration(key encryptionconfig.Key, config encryptionconfig.Config) error {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	if err := key.Validate(); err != nil {
+		return fmt.Errorf("failed to parse decryption fallback configuration from HCL (%w)", err)
+	}
+	meta := encryptionconfig.Meta{
+		encryptionconfig.SourceHCL,
+		key,
+	}
+	e.decryptionFallbackConfigs[meta] = config
+	return nil
+}
+
+func (e *encryption) Validate() (diags tfdiags.Diagnostics) {
+	keys := e.allConfigKeys()
+
+	for key := range keys {
+		if _, err := e.build(key); err != nil {
+			diags = append(diags, tfdiags.Sourceless(
+				tfdiags.Error,
+				fmt.Sprintf("Invalid state encryption configuration for configuration key %s", key),
+				err.Error(),
+			))
+		}
+	}
+
+	return diags
+}
+
+func (e *encryption) allConfigKeys() map[encryptionconfig.Key]struct{} {
```

Comment:
Because a `map[encryptionconfig.Key]struct{}` has an O(1) access time, whereas a slice has a n `O(n)` access time in Go and also guarantees that no duplicate items are present. As ugly as it is, it's the Go way of creating sets.

Reply:
I think that we might even want to get rid of this function as it is called only in one place.

Reply:
Removed and also made sure that `Validate()` works on a single lock.